### PR TITLE
make helper pod configurable 

### PR DESCRIPTION
**Pull request by nicktming**
for #128

### CODE REVIEW

**Comment by mgoltzsche** [comment]
+1

I have similar requirements for [another project](https://github.com/mgoltzsche/jobcachefs) that are met by this PR.

File: main.go [hunk 1]
```diff
@@ -99,6 +101,10 @@ func StartCmd() cli.Command {
 				Usage:  "Required. The ServiceAccountName for deployment",
 				EnvVar: EnvServiceAccountName,
 			},
+			cli.StringFlag{
+				Name:  FlagHelperPodFile,
+				Usage: "Required. The ServiceAccountName for deployment",
+			},
```
Comment (yasker) [inline main.go:107]: Is this required?

File: main.go [hunk 2]
```diff
@@ -99,6 +101,10 @@ func StartCmd() cli.Command {
 				Usage:  "Required. The ServiceAccountName for deployment",
 				EnvVar: EnvServiceAccountName,
 			},
+			cli.StringFlag{
+				Name:  FlagHelperPodFile,
+				Usage: "Required. The ServiceAccountName for deployment",
```
Comment (yasker) [inline main.go:106]: Wrong usage.

File: provisioner.go [hunk 3]
```diff
@@ -10,6 +10,8 @@ import (
 	"sync"
 	"time"
 
+	"strconv"
```
Comment (yasker) [inline provisioner.go:13]: Remove the empty line above.

File: provisioner.go [hunk 4]
```diff
@@ -322,75 +335,66 @@ func (p *LocalPathProvisioner) createHelperPod(action ActionType, cmdsForPath []
...
+
+	helperPod.Namespace = p.namespace
+	helperPod.Spec.NodeName = node
+	helperPod.Spec.ServiceAccountName = p.serviceAccountName
+	helperPod.Spec.RestartPolicy = v1.RestartPolicyNever
+	helperPod.Spec.Tolerations = append(helperPod.Spec.Tolerations, tolerations...)
+	helperPod.Spec.Volumes = append(helperPod.Spec.Volumes, vols...)
+	helperPod.Spec.Containers[0].VolumeMounts = append(helperPod.Spec.Containers[0].VolumeMounts, volMnts...)
+	helperPod.Spec.Containers[0].Command = cmdsForPath
+	helperPod.Spec.Containers[0].Args = []string{filepath.Join(parentDir, volumeDir), strconv.FormatInt(sizeInBytes, 10), volumeMode}
```
Comment (yasker) [inline provisioner.go:397]: We should add a prefix for each parameter now, it's getting confusing.

File: util.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+)
+
+func loadFile(filepath string) (string, error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return "", err
+	}
+	helperPodYaml, err := ioutil.ReadAll(f)
```
Comment (yasker) [inline util.go:18]: `defer` should be here.

**Review by yasker — CHANGES_REQUESTED** [review]
Oh yeah, this approach is much better!

File: main.go (hunk 1 above)
Comment (nicktming) [reply main.go:110]: i add a default value for this flag.

File: main.go (hunk 2 above)
Comment (nicktming) [reply main.go]: ok, changed.

File: provisioner.go (hunk 3 above)
Comment (nicktming) [reply provisioner.go]: ok, done.

File: util.go (hunk 5 above)
Comment (nicktming) [reply util.go:19]: ok, done.

File: provisioner.go (hunk 4 above)
Comment (nicktming) [reply provisioner.go:396]: i add a prefix for these parameters. @yasker could you check for this?

Comment (yasker) [reply provisioner.go:396]: I meant, we can have the script accept parameters by name rather than just order, e.g.

`./setup --path <path> --size <size> --mode <mode>...`

Or we can just do 
`./setup -p <path> -s <size> -m <mode>...`

So the arguments are not solely decided by the order, make it easier to skip a few argument e.g. `size` if not needed, as well as make it clear to the caller how should it use the script.

**Comment by yasker** [comment]
Also, use `rebase master` instead of `merge master` when using git.

File: provisioner.go (hunk 4 above)
Comment (nicktming) [reply provisioner.go]: ok, good idea, it is done now.

**Comment by nicktming** [comment]
> Also, use `rebase master` instead of `merge master` when using git.

@yasker ok, done.

File: provisioner.go [hunk 6]
```diff
@@ -322,75 +334,68 @@ func (p *LocalPathProvisioner) createHelperPod(action ActionType, cmdsForPath []
...
 				},
 			},
 		},
 	}
+	lpvVolumeMounts := []v1.VolumeMount{
+		{
+			Name:      "data",
+			ReadOnly:  false,
+			MountPath: parentDir,
+		},
```
Comment (mgoltzsche) [inline provisioner.go:374]: A CLI option to specify the `MountPropagation` would also be great here.

**Review by mgoltzsche — CHANGES_REQUESTED** [review]

File: provisioner.go (hunk 6 above)
Comment (nicktming) [reply provisioner.go:374]: @yasker what do you think?

Comment (mgoltzsche) [reply provisioner.go:374]: Alternatively, following the template approach, the Pod template could specify the volume mount as well and the code here  could therefore only add the hardcoded volume mount when it is not specified.

Comment (mgoltzsche) [reply provisioner.go:374]: I don't want to block this PR. So if you'd merge it as is I'll create another PR in which I'd also add a corresponding usage example to clarify the feature's purpose.

Comment (nicktming) [reply provisioner.go:374]: > I don't want to block this PR. So if you'd merge it as is I'll create another PR in which I'd also add a corresponding usage example to clarify the feature's purpose.

i will add this flag soon, and  @yasker what do you think?

Comment (nicktming) [reply provisioner.go]: > Alternatively, following the template approach, the Pod template could specify the volume mount as well which is then merged with/over the default, hard-coded volume mount above.
> 
> _It would help my use case already if you'd simply set `MountPropagation=v1.MountPropagationBidirectional` but I am not sure whether this could cause issues with other helpers and the default helper doesn't need it._

i add nodePathType and nodePathMountPropagation flag to enable this feature, you can check it. and @yasker what do you think?

File: main.go [hunk 7]
```diff
@@ -21,23 +22,27 @@ import (
...
+	FlagServiceAccountName       = "service-account-name"
+	DefaultServiceAccount        = "local-path-provisioner-service-account"
+	EnvServiceAccountName        = "SERVICE_ACCOUNT_NAME"
+	FlagKubeconfig               = "kubeconfig"
+	DefaultConfigFileKey         = "config.json"
+	DefaultConfigMapName         = "local-path-config"
+	FlagConfigMapName            = "configmap-name"
+	FlagHelperPodFile            = "helper-pod-file"
+	DefaultHelperPodFile         = "helperPod.yaml"
+	FlagNodePathType             = "node-path-type"
```
Comment (yasker) [inline main.go:44]: Why do we need nodepathType?

File: main.go [hunk 8]
```diff
@@ -101,6 +106,21 @@ func StartCmd() cli.Command {
...
+				Value: "",
+			},
+			cli.StringFlag{
+				Name:  FlagNodePathType,
+				Usage: "Paths to the Helper pod yaml file",
+				Value: string(v1.HostPathDirectoryOrCreate),
+			},
+			cli.StringFlag{
+				Name:  FlagNodePathMountPropagation,
+				Usage: "Paths to the Helper pod yaml file",
```
Comment (yasker) [inline main.go:121]: Wrong usage. And this should be another PR.

Also, I am not sure why we need mount propagation for the helper pod. @mgoltzsche can you explain your use case?

**Review by yasker — CHANGES_REQUESTED** [review]
Sorry for the late review, was busy with Longhorn.

File: main.go (hunk 7 above)
Comment (nicktming) [reply main.go:44]: since @mgoltzsche needs to open mount propagation flag, i made the node-path-type flag open to enable users to decide their own node path type such as BlockDevice, Socket etc.

File: main.go (hunk 8 above)
Comment (mgoltzsche) [reply main.go:121]: @yasker my use case is [this](https://github.com/mgoltzsche/cache-provisioner): provision a cache as container image by running buildah within the helper pod to provision (and commit on PV deletion) the PV directory using overlayfs. However the overlayfs mount must be propagated from the helper pod to the host so that its contents can be mounted as volume to another pod after the helper pod terminated.

Comment (nicktming) [reply main.go]: @yasker @mgoltzsche  i removed mount propagation and node path type flag in this pr, after this pr is done, i will add these flags in next pr.

File: main.go (hunk 7 above)
Comment (nicktming) [reply main.go]: i removed this flag in this pr.

File: main.go (hunk 8 above)
Comment (mgoltzsche) [reply main.go]: @nicktming I'd do the next PR since other requirements came up and I could provide a corresponding usage example as well.

**Review by yasker — APPROVED** [review]
